time series: Parallel Queries + Benchmark

mrtracy opened the pull request:
Allows the server to run the individual queries in an incoming
QueryRequest in parallel. This is a significant performance improvement over the
previous design, which executed the individual queries serially; typical UI
query requests routinely request data for more than 50-100 series at once.

Query worker count is managed via the stoppers "StartLimitedAsyncTask", which
limits the total count of workers for the entire server across all queries. The
limit has been arbitrarily set at 250, but can be adjusted for tests via a
TestingKnob (also added in this commit).

Also includes a commit to add a benchmark test for multi-query requests. This is added in a preceding commit, in order to demonstrate the difference in performance from the old serial execution. Benchstat output (50 runs):

```
name           old time/op  new time/op  delta
ServerQuery-8  9.76ms ± 9%  4.11ms ±13%  -57.91%  (p=0.000 n=50+50)
```

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10250)

<!-- Reviewable:end -->

Commit:
Time Series: Benchmark for large Queries

Creates a benchmark test which measures the performance of time series queries
that have multiple subqueries. The number of queries in the request has been set
to 50, which is typical of current usage by the UI.

This is intended to demonstrate the benefit of a subsequent commit, which will
allow queries to execute in parallel.

Commit:
Time Series: Parallel Queries

This commit allows the server to run the individual queries in an incoming
QueryRequest in parallel. This is a significant performance improvement over the
previous design, which executed the individual queries serially; typical UI
query requests routinely request data for more than 50-100 series at once.

Query worker count is managed via the stoppers "StartLimitedAsyncTask", which
limits the total count of workers for the entire server across all queries. The
limit has been set to a default of 250, but can be adjusted via a configuration
value passed to the time series server.

tschottdorf commented:
I'm curious - shouldn't parallel `DistSender` obviate parallel execution here?

mrtracy commented:
No, what needs to be called in parallel is the `tsdb.Query()` method: it needs to be called once for each query in an incoming request.  The `Query()` method itself calls distsender to look up data; it is also responsible for performing all of our time series downsampling/aggregation/interpolation magic on the data it finds.

petermattis commented:
> No, what needs to be called in parallel is the tsdb.Query() method: it needs to be called once for each query in an incoming request. 

Isn't that due to the interface to `tsdb.Query`? I could imagine a different API where `tsdb.Query` populated a `client.Batch` and a separate method processed the results from it. Certainly that would be a more significant refactoring than this PR.

mrtracy commented:
That would be a considerable refactoring, and I don't think it would even be better; unless results from DistSender are streaming, we would have to complete the query of all data before we begin processing *any* data, whereas this method can begin processing data for each query as soon as data is available.

petermattis commented on pkg/ts/server.go:
```diff
@@ -22,30 +22,57 @@ import (
 	URLPrefix = "/ts/"
+	// QueryWorkerMax determines the maximum number of worker goroutines that
+	// the time series server can use to service incoming queries.
+	QueryWorkerMax = 250
```
I think you can unexport this.

petermattis commented on pkg/ts/server.go:
```diff
@@ -22,30 +22,57 @@ import (
+// test server that are not normally overridden.
+type TimeSeriesTestingKnobs struct {
+	// Adjusts the maximum number of query workers used by the server.
+	QueryWorkerMax *int
```
Making this a pointer is slightly awkward. Elsewhere we use `0` as an indicator that a value should assume the default. For example, see `StoreConfig.SetDefaults`.

petermattis commented on pkg/ts/server.go:
```diff
@@ -78,28 +105,105 @@ func (s *Server) Query(
+	// they have written their result to the "output" channel, which is
+	// processed later in the main function.
+	if err := s.stopper.RunAsyncTask(ctx, func(loopCtx context.Context) {
+		for queryNum := 0; queryNum < len(request.Queries); queryNum++ {
```
`for queryNum := range request.Queries`?

tamird commented:
Reviewed 1 of 1 files at r1, 3 of 4 files at r2.
Review status: 3 of 4 files reviewed at latest revision, 10 unresolved discussions.

---

*[pkg/ts/server.go, line 43 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV2ZiEddiodN_63ieUE:-KV2ZiEddiodN_63ieUF:bj8nfi7) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9164c87770a0a43ea0e6aa284238b4e7807bd8/pkg/ts/server.go#L43)):*
> ```Go
> // TimeSeriesTestingKnobs provides a means for tests to override settings in the
> // test server that are not normally overridden.
> type TimeSeriesTestingKnobs struct {
> ```

I think this is getting ahead of yourself, and also it's not really appropriate for this to be a "testing knob". It's really more akin to a config object, like storage.StoreConfig.

For now, you can just pass an integer to MakeServer, and if that integer is zero, use a default.

---

*[pkg/ts/server.go, line 112 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV2bI_rZGbgv5cbyam5:-KV2Zz66CZoe4RfHMGX4:b-2ndw6t) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9164c87770a0a43ea0e6aa284238b4e7807bd8/pkg/ts/server.go#L112)):*
> ```Go
> 	// in the same order as the queries in the request.
> 	type queryResultWithError struct {
> 		queryNum int
> ```

s/Num/Idx/g

---

*[pkg/ts/server.go, line 129 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV2bjODJ1jTHTog3rzB:-KV2bjODJ1jTHTog3rzC:bj0p6df) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9164c87770a0a43ea0e6aa284238b4e7807bd8/pkg/ts/server.go#L129)):*
> ```Go
> 	// they have written their result to the "output" channel, which is
> 	// processed later in the main function.
> 	if err := s.stopper.RunAsyncTask(ctx, func(loopCtx context.Context) {
> ```

i think you actually do want to shadow `ctx` in all these closures; each closure should only be able to access the innermost context.

---

*[pkg/ts/server.go, line 130 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV2bKliXPf4WSVIsmnH:-KV2a6sx5kYiMQCLZsBP:b-y2oda3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9164c87770a0a43ea0e6aa284238b4e7807bd8/pkg/ts/server.go#L130)):*
> ```Go
> 	// processed later in the main function.
> 	if err := s.stopper.RunAsyncTask(ctx, func(loopCtx context.Context) {
> 		for queryNum := 0; queryNum < len(request.Queries); queryNum++ {
> ```

`for queryIdx, query := range request.Queries` ?

---

*[pkg/ts/server.go, line 140 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV2asIuDpHcQPSVp4ka:-KV2asIvByOCh67Ypt-S:b-vel09i) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9164c87770a0a43ea0e6aa284238b4e7807bd8/pkg/ts/server.go#L140)):*
> ```Go
> 
> 			queryNum := queryNum
> 			if err := s.stopper.RunLimitedAsyncTask(
> ```

this can block for some time even after the context has been cancelled, can't it?

I think you should enhance `Run{,Limited}AsyncTask` to watch for cancellation on the context it's provided, which will allow you to omit the `select` above and will actually behave more as you intend.

---

*[pkg/ts/server.go, line 192 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV2agGqm9EuN92DspBP:-KV2agGqm9EuN92DspBQ:b-r0rtef) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9164c87770a0a43ea0e6aa284238b4e7807bd8/pkg/ts/server.go#L192)):*
> ```Go
> 	}
> 
> 	completed := 0
> ```

move into the loop

---

*[pkg/ts/server.go, line 203 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV2ak25_Xal940D0BW2:-KV2ak25_Xal940D0BW3:bbxjote) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9164c87770a0a43ea0e6aa284238b4e7807bd8/pkg/ts/server.go#L203)):*
> ```Go
> 			}
> 			response.Results[queryResult.queryNum] = queryResult.result
> 			completed++
> ```

move into the `for` line

---

*[pkg/ts/server_test.go, line 265 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV2YbK6W-14qFFGg4LU:-KV2YbK7iQS9FnUwJOrE:b40lr9z) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd5135d1e7584fc2ba14acc39d6f54cb75e868b1/pkg/ts/server_test.go#L265)):*
> ```Go
> 	b.ResetTimer()
> 
> 	queries := make([]tspb.Query, 0, seriesCount)
> ```

why include the construction of this slice in the benchmark time?

---

*[pkg/ts/server_test.go, line 275 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV2YgSN_LXK1XXXVwkU:-KV2YgSN_LXK1XXXVwkV:b321e9v) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd5135d1e7584fc2ba14acc39d6f54cb75e868b1/pkg/ts/server_test.go#L275)):*
> ```Go
> 		if _, err := client.Query(context.Background(), &tspb.TimeSeriesQueryRequest{
> 			StartNanos: 0 * 1e9,
> 			EndNanos:   500 * 1e9,
> ```

math.MaxInt64? or whatever the correct type.

---

*[pkg/ts/server_test.go, line 291 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV2YrGRmVZMaxo5EQF4:-KV2YrGRmVZMaxo5EQF5:bisx3tb) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd5135d1e7584fc2ba14acc39d6f54cb75e868b1/pkg/ts/server_test.go#L291)):*
> ```Go
> }
> 
> func populateSeries(t testing.TB, seriesCount, sourceCount int, tsdb *ts.DB) {
> ```

please return an error instead of Fataling inside.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250)*
<!-- Sent from Reviewable.io -->

petermattis commented on pkg/ts/server.go:
```diff
@@ -78,28 +105,105 @@ func (s *Server) Query(
+	}
+
+	completed := 0
+	for completed < len(request.Queries) {
```
`for range request.Queries`. If the `ctx.Done()` case occurs you're returning and don't care about the index.

Commit:
util: stopper.RunLimitedAsyncTask honors context

Added logic to stopper.RunLimitedAsyncTask in order to respect cancelation of
the supplied context. Previously, RunLimitedAsyncTask would stop blocking (in
the case where the semaphore is full) if the stopper itself is quiesced; now
it will also stop if the required context is canceled.

Includes a test to verify this behavior.

mrtracy commented:
Review status: 3 of 4 files reviewed at latest revision, 15 unresolved discussions.

---

*[pkg/ts/server.go, line 38 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV1tO-o7U23uVoAWOdo-r2-38:-KV6sWxORQL10aH72Qwh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9164c87770a0a43ea0e6aa284238b4e7807bd8/pkg/ts/server.go#L38)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I think you can unexport this.
</details>
Done.

---

*[pkg/ts/server.go, line 43 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV2ZiEddiodN_63ieUE:-KV6sXRBupO69mr9xZyn:b-a9vix2) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9164c87770a0a43ea0e6aa284238b4e7807bd8/pkg/ts/server.go#L43)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I think this is getting ahead of yourself, and also it's not really appropriate for this to be a "testing knob". It's really more akin to a config object, like storage.StoreConfig.
> 
> For now, you can just pass an integer to MakeServer, and if that integer is zero, use a default.
</details>
I did that initially; my thought is, that this is the right thing to do and resembles what is done for other subsystems similar to TimeSeries.  This value is not being configured by anything other than a test - it's a testing knob, and I think it should be treated as self.

I don't think it adds much complexity, and will make the next attempt to add test-related settings more straightforward.

---

*[pkg/ts/server.go, line 45 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV1tO-o7U23uVoAWOdo-r2-45:-KV6t5BJ5jjw1Q6GXFEh:b-avj0ty) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9164c87770a0a43ea0e6aa284238b4e7807bd8/pkg/ts/server.go#L45)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Making this a pointer is slightly awkward. Elsewhere we use `0` as an indicator that a value should assume the default. For example, see `StoreConfig.SetDefaults`.
</details>
Done.  I chose the wrong pattern because of `StoreTestingKnob.NumKeysEvaluatedForRangeIntentResolution`, which *is* set to an integer pointer, but is apparently significantly more complex and *not* used as a simple configuration value.

---

*[pkg/ts/server.go, line 112 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV2bI_rZGbgv5cbyam5:-KV6tv9pTc3UhtXBe_bL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9164c87770a0a43ea0e6aa284238b4e7807bd8/pkg/ts/server.go#L112)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/Num/Idx/g
</details>
Done.

---

*[pkg/ts/server.go, line 129 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV2bjODJ1jTHTog3rzB:-KV6u3Jr8Kwh1_nYfhrq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9164c87770a0a43ea0e6aa284238b4e7807bd8/pkg/ts/server.go#L129)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> i think you actually do want to shadow `ctx` in all these closures; each closure should only be able to access the innermost context.
</details>
Done.

---

*[pkg/ts/server.go, line 130 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV1tO-o7U23uVoAWOdo-r2-130:-KV6uCh4iYihzL6jXKME:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9164c87770a0a43ea0e6aa284238b4e7807bd8/pkg/ts/server.go#L130)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `for queryNum := range request.Queries`?
</details>
Done.

---

*[pkg/ts/server.go, line 130 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV2bKliXPf4WSVIsmnH:-KV6uIaCZEJmiljYA-Pa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9164c87770a0a43ea0e6aa284238b4e7807bd8/pkg/ts/server.go#L130)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `for queryIdx, query := range request.Queries` ?
</details>
Done.

---

*[pkg/ts/server.go, line 140 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV2asIuDpHcQPSVp4ka:-KV723hV0o2CIff8OMlA:b-u6foaq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9164c87770a0a43ea0e6aa284238b4e7807bd8/pkg/ts/server.go#L140)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this can block for some time even after the context has been cancelled, can't it?
> 
> I think you should enhance `Run{,Limited}AsyncTask` to watch for cancellation on the context it's provided, which will allow you to omit the `select` above and will actually behave more as you intend.
</details>
I thought about doing this, your comment convinced me to do it. It's only being used in a small number of places, and they do not appear to be sensitive to this; I would argue that honoring the context should be the expected behavior, if we're truly committed to using contexts everywhere.

Augmented as a separate commit, with an additional test.

---

*[pkg/ts/server.go, line 192 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV2agGqm9EuN92DspBP:-KV72Z-e84xUQvGNqaQU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9164c87770a0a43ea0e6aa284238b4e7807bd8/pkg/ts/server.go#L192)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> move into the loop
</details>
Done.

---

*[pkg/ts/server.go, line 193 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV1tO-o7U23uVoAWOdo-r2-193:-KV734mbGtZTDGl8mE4O:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9164c87770a0a43ea0e6aa284238b4e7807bd8/pkg/ts/server.go#L193)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `for range request.Queries`. If the `ctx.Done()` case occurs you're returning and don't care about the index.
</details>
Done.

---

*[pkg/ts/server.go, line 203 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV2ak25_Xal940D0BW2:-KV735CBEneMrIlz0cyA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9164c87770a0a43ea0e6aa284238b4e7807bd8/pkg/ts/server.go#L203)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> move into the `for` line
</details>
Done.

---

*[pkg/ts/server_test.go, line 265 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV2YbK6W-14qFFGg4LU:-KV73utN3Y1NjkG0pSnI:b6g3atp) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd5135d1e7584fc2ba14acc39d6f54cb75e868b1/pkg/ts/server_test.go#L265)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why include the construction of this slice in the benchmark time?
</details>
Just a total oversight, fixed.

---

*[pkg/ts/server_test.go, line 275 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV2YgSN_LXK1XXXVwkU:-KV73u8QU_TbQG_jdcLi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd5135d1e7584fc2ba14acc39d6f54cb75e868b1/pkg/ts/server_test.go#L275)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> math.MaxInt64? or whatever the correct type.
</details>
Done.

---

*[pkg/ts/server_test.go, line 291 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV2YrGRmVZMaxo5EQF4:-KV73uQ6ABrJjYyVrddD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd5135d1e7584fc2ba14acc39d6f54cb75e868b1/pkg/ts/server_test.go#L291)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> please return an error instead of Fataling inside.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 0 of 6 files reviewed at latest revision, 15 unresolved discussions, some commit checks pending.

---

*[pkg/ts/server.go, line 128 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV1tO-o7U23uVoAWOdo-r2-128:-KV7HEg4yMUqDmRIwUpD:b94ikxd) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9164c87770a0a43ea0e6aa284238b4e7807bd8/pkg/ts/server.go#L128)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary>
> I think this will be less awkward if you always create the goroutine as you do below and then use a semaphore within it rather than calling RunLimitedAsyncTask. Perhaps you can get rid of one of the selects below
</details>
I augmented RunLimitedAsyncTask to handle context cancelation correctly, which got rid of one of the select statements.  Other than that, it's basically doing exactly what I want already, the only tricky part is avoiding deadlocks with the semaphore - that can occur in the case where one Request has more queries than the semaphore's max value. 

I have written a test to verify that this situation is safe.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 6 files reviewed at latest revision, 13 unresolved discussions, all commit checks successful.

---

*[pkg/ts/server.go, line 36 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV7gqPKH6_qmgADj1Do:-KV7gqPKH6_qmgADj1Dp:b-7oyfo0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3aee2d354dfe8d987396e5e79eb935f9cd18421/pkg/ts/server.go#L36)):*
> ```Go
> 	// server.
> 	URLPrefix = "/ts/"
> 	// QueryWorkerMax determines the maximum number of worker goroutines that
> ```

`s/QueryWorkerMax/queryWorkerMax/g`

---

*[pkg/ts/server.go, line 184 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV7hKFcj3UW5-CVktFo:-KV7hKFcj3UW5-CVktFp:bmq41ds) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3aee2d354dfe8d987396e5e79eb935f9cd18421/pkg/ts/server.go#L184)):*
> ```Go
> 	response := tspb.TimeSeriesQueryResponse{
> 		Results: make([]tspb.TimeSeriesQueryResponse_Result, len(request.Queries)),
> 	}
> ```

I could have sworn I wrote a comment here, but it disappeared.

If you move this variable above your loop, you can have each async task write the result directly to `response.Results[queryNum]` which then simplifies your `workerOutput` channel to type `chan error`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-:-KV7iFtXFjuW6pRexvFv:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 6 of 6 files at r3, 1 of 1 files at r4, 4 of 4 files at r5.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/ts/server.go, line 43 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV2ZiEddiodN_63ieUE:-KVAm1tWDkUz-MjS05ry:bizfmr) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9164c87770a0a43ea0e6aa284238b4e7807bd8/pkg/ts/server.go#L43)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> I did that initially; my thought is, that this is the right thing to do and resembles what is done for other subsystems similar to TimeSeries.  This value is not being configured by anything other than a test - it's a testing knob, and I think it should be treated as self.
> 
> I don't think it adds much complexity, and will make the next attempt to add test-related settings more straightforward.
</details>
It's certainly not a testing knob; it's a general configuration of the object. A testing knob is typically a callback or something that is expected to be unused outside of tests. This is just a concurrency value, and these are treated as configuration in every single other component.

---

*[pkg/ts/server.go, line 130 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVAo4_wPKwVt-7haijs:-KVAo4_wPKwVt-7haijt:b-h6ftq2) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3aee2d354dfe8d987396e5e79eb935f9cd18421/pkg/ts/server.go#L130)):*
> ```Go
> 	// processed later in the main function.
> 	if err := s.stopper.RunAsyncTask(ctx, func(ctx context.Context) {
> 		for queryNum, query := range request.Queries {
> ```

again, s/Num/Idx/

---

*[pkg/ts/server_test.go, line 316 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVAnneBWfN0bfub1K7T:-KVAnneBWfN0bfub1K7U:b-86oyy3) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9a00e018170cd16220f5c5ec3b5524e68f4084/pkg/ts/server_test.go#L316)):*
> ```Go
> 				},
> 			}); err != nil {
> 				return fmt.Errorf(
> ```

 it: errors.Errorf

---

*[pkg/util/stop/stopper_test.go, line 483 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVAn6HXczj1214lydv_:-KVAn6HYASj-Macqf0XN:b-qfu2da) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e5af7880c4f8796d3d597bf49047db3aa01c54e/pkg/util/stop/stopper_test.go#L483)):*
> ```Go
> 	workerStarted := make(chan struct{})
> 
> 	var workersRun int32 = 0
> ```

nit: no need to `= 0` and we also don't use that style anywhere?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 1 of 10 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[pkg/ts/server.go, line 43 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV2ZiEddiodN_63ieUE:-KVBjJN4wXB39zbD9iA2:buwu6sy) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e9164c87770a0a43ea0e6aa284238b4e7807bd8/pkg/ts/server.go#L43)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> It's certainly not a testing knob; it's a general configuration of the object. A testing knob is typically a callback or something that is expected to be unused outside of tests. This is just a concurrency value, and these are treated as configuration in every single other component.
</details>
Changed this structure to "ServerConfig", removed it from TestingKnobs and added it to the main server config.

---

*[pkg/ts/server.go, line 36 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV7gqPKH6_qmgADj1Do:-KVBafjKrBUlCYk0ATq5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3aee2d354dfe8d987396e5e79eb935f9cd18421/pkg/ts/server.go#L36)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/QueryWorkerMax/queryWorkerMax/g`
</details>
Done.

---

*[pkg/ts/server.go, line 130 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVAo4_wPKwVt-7haijs:-KVBiydhJ311o7Cl23S6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3aee2d354dfe8d987396e5e79eb935f9cd18421/pkg/ts/server.go#L130)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> again, s/Num/Idx/
</details>
Done.

---

*[pkg/ts/server.go, line 184 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KV7hKFcj3UW5-CVktFo:-KVBjA4wzh6sJMsuCjUB:bbsxbaq) ([raw file](https://github.com/cockroachdb/cockroach/blob/e3aee2d354dfe8d987396e5e79eb935f9cd18421/pkg/ts/server.go#L184)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I could have sworn I wrote a comment here, but it disappeared.
> 
> If you move this variable above your loop, you can have each async task write the result directly to `response.Results[queryNum]` which then simplifies your `workerOutput` channel to type `chan error`. 
</details>
That was my original thought, I didn't think that design would make it through the race detector without a mutex. Apparently, the race detector is better than I give it credit for.  Changed.

---

*[pkg/ts/server_test.go, line 316 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVAnneBWfN0bfub1K7T:-KVBdWNmSNmCVmoIoe2Q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9a00e018170cd16220f5c5ec3b5524e68f4084/pkg/ts/server_test.go#L316)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
>  it: errors.Errorf
</details>
Done.

---

*[pkg/util/stop/stopper_test.go, line 483 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVAn6HXczj1214lydv_:-KVBawEL8YEaPQbEzQj7:b20c3t6) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e5af7880c4f8796d3d597bf49047db3aa01c54e/pkg/util/stop/stopper_test.go#L483)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: no need to `= 0` and we also don't use that style anywhere?
</details>
What style? The `=0`?  Removed the `=0`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 5 files at r6, 3 of 3 files at r7, 6 of 6 files at r8.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/ts/new.txt, line 1 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVBoPc_yW818hy3L2Db:-KVBoPc_yW818hy3L2Dc:b-z2x071) ([raw file](https://github.com/cockroachdb/cockroach/blob/517e6138a1f83735b8c46ddbf22e112ab370c0a7/pkg/ts/new.txt#L1)):*
> ```
> FAIL	github.com/cockroachdb/cockroach/pkg/ts [build failed]
> ```

remove these files

---

*[pkg/ts/old.txt, line 1 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVBoRpBfks_TivYezQt:-KVBoRpBfks_TivYezQu:b-z2x071) ([raw file](https://github.com/cockroachdb/cockroach/blob/517e6138a1f83735b8c46ddbf22e112ab370c0a7/pkg/ts/old.txt#L1)):*
> ```
> FAIL	github.com/cockroachdb/cockroach/pkg/ts [build failed]
> ```

remove these files

---

*[pkg/ts/server.go, line 43 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVBogU2FSSMguVkH66f:-KVBogU2FSSMguVkH66g:bvybb21) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1f9e6bec727ae40dcc5e26ea27200401262a0e3/pkg/ts/server.go#L43)):*
> ```Go
> // series server.
> type ServerConfig struct {
> 	// Adjusts the maximum number of query workers used by the server. If this
> ```

s/Adjusts the/The/

---

*[pkg/ts/server_test.go, line 316 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVAnneBWfN0bfub1K7T:-KVBowUclyG82PXfFYOs:bmyv3n4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9a00e018170cd16220f5c5ec3b5524e68f4084/pkg/ts/server_test.go#L316)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> Done.
</details>
wrong commit, though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-:-KVBp4ggFmMGRpKKLnJI:bnfp4nl)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/base/test_server_args.go, line 61 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVBsV2xab-nZZVwEYvZ:-KVBsV2xab-nZZVwEYv_:bgyf0gl) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1f9e6bec727ae40dcc5e26ea27200401262a0e3/pkg/base/test_server_args.go#L61)):*
> ```Go
> 	SSLCert                  string
> 	SSLCertKey               string
> 	TimeSeriesQueryWorkerMax int
> ```

This is really preferable to a `TestingKnob`? Previously this knob was local to the `ts` package, and now it has snuck into `base`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/base/test_server_args.go, line 61 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVBsV2xab-nZZVwEYvZ:-KVClPRyG92w2ZI8p8sf:bkcgdsz) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1f9e6bec727ae40dcc5e26ea27200401262a0e3/pkg/base/test_server_args.go#L61)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> This is really preferable to a `TestingKnob`? Previously this knob was local to the `ts` package, and now it has snuck into `base`.
</details>
There was code in base previously, on the "TestingKnobs" struct; this is analogous.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/base/test_server_args.go, line 61 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVBsV2xab-nZZVwEYvZ:-KVCpd2khbdb--piIYja:b-nblswd) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1f9e6bec727ae40dcc5e26ea27200401262a0e3/pkg/base/test_server_args.go#L61)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> There was code in base previously, on the "TestingKnobs" struct; this is analogous.
</details>
The previous timeseries `TestingKnobs` hook was general purpose and allowed the details of the testing knobs to be left in the `ts` package. This field is specific. 

PS This field is not copied to `server.Config`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/base/test_server_args.go, line 61 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVBsV2xab-nZZVwEYvZ:-KVCqKWB6LZCKj2EW2kZ:b-lx5hwh) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1f9e6bec727ae40dcc5e26ea27200401262a0e3/pkg/base/test_server_args.go#L61)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> The previous timeseries `TestingKnobs` hook was general purpose and allowed the details of the testing knobs to be left in the `ts` package. This field is specific. 
> 
> PS This field is not copied to `server.Config`.
</details>
In my opinion, this field was perfectly fine as a `TestingKnob`. @tamird Can you expand on why you objected to that usage? I'm not seeing any comment in `base/testing_knobs.go` which would indicate that this field doesn't belong in one. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/base/test_server_args.go, line 61 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVBsV2xab-nZZVwEYvZ:-KVQdWQJxVKBxoEN9azk:b-mbxv4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1f9e6bec727ae40dcc5e26ea27200401262a0e3/pkg/base/test_server_args.go#L61)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> In my opinion, this field was perfectly fine as a `TestingKnob`. @tamird Can you expand on why you objected to that usage? I'm not seeing any comment in `base/testing_knobs.go` which would indicate that this field doesn't belong in one. 
</details>
Is it fine to check in either way?

---

*[pkg/ts/new.txt, line 1 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVBoPc_yW818hy3L2Db:-KVQdfF_jfFt_dvnPJ00:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/517e6138a1f83735b8c46ddbf22e112ab370c0a7/pkg/ts/new.txt#L1)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> remove these files
</details>
Done.

---

*[pkg/ts/old.txt, line 1 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVBoRpBfks_TivYezQt:-KVQdfjZJ_sUzVma-YHI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/517e6138a1f83735b8c46ddbf22e112ab370c0a7/pkg/ts/old.txt#L1)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> remove these files
</details>
Done.

---

*[pkg/ts/server.go, line 43 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVBogU2FSSMguVkH66f:-KVQdkSqnghquPSyyoZQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1f9e6bec727ae40dcc5e26ea27200401262a0e3/pkg/ts/server.go#L43)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/Adjusts the/The/
</details>
Done.

---

*[pkg/ts/server_test.go, line 316 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVAnneBWfN0bfub1K7T:-KVQePPb9RR_PWNv8rCC:bag1jx6) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9a00e018170cd16220f5c5ec3b5524e68f4084/pkg/ts/server_test.go#L316)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> wrong commit, though.
</details>
I think it will be fine.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 1 of 8 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[pkg/base/test_server_args.go, line 61 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVBsV2xab-nZZVwEYvZ:-KVQgRfSaSUttMp_Ing2:b1t37xl) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1f9e6bec727ae40dcc5e26ea27200401262a0e3/pkg/base/test_server_args.go#L61)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> Is it fine to check in either way?
</details>
And this field *is* copied to server.config, in testserver.go makeTestConfigFromParams().

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 8 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/base/test_server_args.go, line 61 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVBsV2xab-nZZVwEYvZ:-KVQjBBqBKR2JYftU8XG:b-kc5v3u) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1f9e6bec727ae40dcc5e26ea27200401262a0e3/pkg/base/test_server_args.go#L61)):*
Yes, fine to check in.

> And this field is copied to server.config, in testserver.go makeTestConfigFromParams().

Understood. The ugliness is that you're exposing a timeseries parameter in the base package. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 9 files at r9, 6 of 6 files at r11.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/base/test_server_args.go, line 61 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVBsV2xab-nZZVwEYvZ:-KVQmFVtzzhdisrmx-D2:b-ch7dgg) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1f9e6bec727ae40dcc5e26ea27200401262a0e3/pkg/base/test_server_args.go#L61)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Yes, fine to check in.
> 
> > And this field is copied to server.config, in testserver.go makeTestConfigFromParams().
> 
> Understood. The ugliness is that you're exposing a timeseries parameter in the base package. 
</details>
@petermattis clearly you feel more strongly than I. There is precedent for such configuration objects being configuration objects and not testing knobs, though the situation is largely a mess. For example, why do we need configuration of dist sender, when exactly one exists for the life of the program? Just for testing, and yet we don't call its configuration a testing knob.

---

*[pkg/ts/server.go, line 43 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVBogU2FSSMguVkH66f:-KVQmbuXtRZrP22PYGpw:b-slwcil) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1f9e6bec727ae40dcc5e26ea27200401262a0e3/pkg/ts/server.go#L43)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> Done.
</details>
Needs re-wrap?

---

*[pkg/ts/server_test.go, line 316 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVAnneBWfN0bfub1K7T:-KVQmguj2YJSuATzbQ0g:b-2h4r3r) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9a00e018170cd16220f5c5ec3b5524e68f4084/pkg/ts/server_test.go#L316)):*
<details><summary><i>Previously, mrtracy wrote…</i></summary>
> I think it will be fine.
</details>
... you already ate another CI. just objecting to object?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/base/test_server_args.go, line 61 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVBsV2xab-nZZVwEYvZ:-KVQn5POOh5dvPobe8uM:b7qwwsn) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1f9e6bec727ae40dcc5e26ea27200401262a0e3/pkg/base/test_server_args.go#L61)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> @petermattis clearly you feel more strongly than I. There is precedent for such configuration objects being configuration objects and not testing knobs, though the situation is largely a mess. For example, why do we need configuration of dist sender, when exactly one exists for the life of the program? Just for testing, and yet we don't call its configuration a testing knob.
</details>
I don't feel that strongly as I'm ok with this going in as-is. I agree that the `*Config` and `*TestingKnob` distinction isn't always clear. Perhaps we need a mechanism for hooking per-component `*Config` structures into `Test{Server,Cluster}` similar to what we have for `*TestingKnobs`. This isn't a high priority, though.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/ts/server_test.go, line 316 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250#-KVAnneBWfN0bfub1K7T:-KVQr74nsOHHNQOA0HGS:b-9iu956) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a9a00e018170cd16220f5c5ec3b5524e68f4084/pkg/ts/server_test.go#L316)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ... you already ate another CI. just objecting to object?
</details>
Objecting because the fix was itself a nit; "previous nit was addressed in the wrong commit of this multi-commit PR" is on a different level of nit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10250)*
<!-- Sent from Reviewable.io -->
